src/methods/join.js

```diff
@@ -99,405 +91,128 @@ function joinArraysOfValues(
-      }
-      invariant(pncc instanceof PossiblyNormalCompletion);
-      let nc = recurse(pncc, pnc.consequentEffects, ac, e);
-      return new ForkedAbruptCompletion(realm, pnc.joinCondition, nc, pnca);
-    }
-
-    // * case (SimpleNormalCompletion, SimpleNormalCompletion)
-    // * case (SimpleNormalCompletion, PossibleNormalCompletion)
-    if (pncc instanceof SimpleNormalCompletion) {
-      let nce = realm.composeEffects(pnc.consequentEffects, e);
-      invariant(nce.result instanceof AbruptCompletion);
-      let nc = nce.result;
-      [ac, e] = cloneEffects();
-      let na, nae;
-      if (pnca instanceof SimpleNormalCompletion) {
-        nae = realm.composeEffects(pnc.alternateEffects, e);
-        invariant(nae.result instanceof AbruptCompletion);
-        na = nae.result;
-      } else {
-        invariant(pnca instanceof PossiblyNormalCompletion);
-        na = recurse(pnca, pnc.alternateEffects, ac, e);
-      }
-      return new ForkedAbruptCompletion(realm, pnc.joinCondition, nc, na);
-    }
-
-    // * case (PossibleNormalCompletion, SimpleNormalCompletion)
-    if (pnca instanceof SimpleNormalCompletion) {
-      let nae = realm.composeEffects(pnc.alternateEffects, e);
-      invariant(nae.result instanceof AbruptCompletion);
-      let na = nae.result;
-      invariant(pncc instanceof PossiblyNormalCompletion);
-      [ac, e] = cloneEffects();
-      let nc = recurse(pncc, pnc.consequentEffects, ac, e);
-      return new ForkedAbruptCompletion(realm, pnc.joinCondition, nc, na);
-    }
-
-    // * case (PossibleNormalCompletion, PossibleNormalCompletion)
-    invariant(pncc instanceof PossiblyNormalCompletion);
-    invariant(pnca instanceof PossiblyNormalCompletion);
-    let nc = recurse(pncc, pnc.consequentEffects, ac, e);
-    [ac, e] = cloneEffects();
-    let na = recurse(pnca, pnc.alternateEffects, ac, e);
-    return new ForkedAbruptCompletion(realm, pnc.joinCondition, nc, na);
-
-    // Impossible cases:
-    // * case (AbruptCompletion, AbruptCompletion)
+  composeWithEffects(completion: Completion, effects: Effects): Effects {
+    if (completion instanceof AbruptCompletion) return construct_empty_effects(completion.value.$Realm, completion);
+    if (completion instanceof SimpleNormalCompletion) return effects.shallowCloneWithResult(effects.result);
+    invariant(completion instanceof JoinedNormalAndAbruptCompletions);
+    let e1 = this.composeWithEffects(completion.consequent, effects);
+    let e2 = this.composeWithEffects(completion.alternate, effects);
+    return this.joinEffects(completion.joinCondition, e1, e2);
   }
 
-  joinNormalCompletions(
-    realm: Realm,
-    joinCondition: AbstractValue,
-    c: NormalCompletion,
-    ce: Effects,
-    a: NormalCompletion,
-    ae: Effects
-  ): PossiblyNormalCompletion {
+  _collapseSimilarCompletions(joinCondition: AbstractValue, c1: Completion, c2: Completion): void | Completion {
```

Comment:
It is a fair question, but I don't know the answer. It might be an improvement, but it is not entirely obvious to me that it would be. For now, it is probably not the most important thing we have to worry about.